### PR TITLE
[youtube] add invidious-redirect websites

### PR DESCRIPTION
**Pull request by B0pol**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x[ Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Add invidious-redirect websites, i.e. websites running [invidious-redirect](https://github.com/iv-org/invidious-redirect), they redirect to all public invidious instances and then are often used to share URLs

### CODE REVIEW

File: youtube_dl/extractor/youtube.py [hunk 1]
```diff
@@ -335,6 +335,9 @@ class YoutubeIE(YoutubeBaseInfoExtractor):
                             (?:www\.)?hooktube\.com/|
                             (?:www\.)?yourepeat\.com/|
                             tube\.majestyc\.net/|
+                            # Invidious redirect websites
+                            (?:www\.)?invidio\.us/|
```
Comment (dstftw) [inline youtube_dl/extractor/youtube.py:339]: This is already matched.

File: youtube_dl/extractor/youtube.py [hunk 2]
```diff
@@ -335,6 +335,9 @@ class YoutubeIE(YoutubeBaseInfoExtractor):
...
+                            (?:www\.)?redirect\.invidious\.io/|
```
Comment (dstftw) [inline youtube_dl/extractor/youtube.py:340]: Add a test.

**Review by dstftw — CHANGES_REQUESTED** [review]

File: youtube_dl/extractor/youtube.py (hunk 2 above)
Comment (B0pol) [reply youtube_dl/extractor/youtube.py:339]: done

File: youtube_dl/extractor/youtube.py [hunk 3]
```diff
@@ -906,6 +908,10 @@ class YoutubeIE(YoutubeBaseInfoExtractor):
             'url': 'https://invidio.us/watch?v=BaW_jenozKc',
             'only_matching': True,
         },
+        {
+            'url': 'https://redirect.invidious.io/watch?v=BaW_jenozKc',
```
Comment (dstftw) [inline youtube_dl/extractor/youtube.py:912]: There is no video on this page.

Comment (B0pol) [reply youtube_dl/extractor/youtube.py:912]: so is on the test above, which is invidio.us

please read the PR description

> Add invidious-redirect websites, i.e. websites running invidious-redirect, they redirect to all public invidious instances and then are often used to share URLs

and then (maybe) please read [invidious-redirect's README](https://github.com/iv-org/invidious-redirect)

Comment (dstftw) [reply youtube_dl/extractor/youtube.py:912]: At the moment of writing, https://invidio.us/watch?v=BaW_jenozKc had BaW_jenozKc embedded so this patterns is kept for compatibility with old URLs possibly still floating over the net. https://redirect.invidious.io/watch?v=BaW_jenozKc does not embed any video and apparently never did thus has nothing to do in this case.

Comment (B0pol) [reply youtube_dl/extractor/youtube.py:912]: ok then what about youtu.be? there is no video on youtu.be, it redirects to youtube.com/
same with redirect.invidious.io

even if there is no video the url path on redirect.invidious.io is right now invidious's same path and it **is** used for URLs floating over the net to share youtube videos.

Comment (dstftw) [reply youtube_dl/extractor/youtube.py:912]: This does not matter, if you open https://youtu.be/BaW_jenozKc there is a video embedded.
If you open https://redirect.invidious.io/watch?v=BaW_jenozKc there is no video embedded.
>used for URLs floating over the net to share youtube videos.

Prove that.

Comment (B0pol) [reply youtube_dl/extractor/youtube.py:912]: > Prove that.

some software use invidious-redirect (redirect.invidious.io) as YouTube URL rewrite:
- [bibliogram](https://git.sr.ht/~cadence/bibliogram/tree/master/item/src/lib/constants.js#L111)
- [cloudtube](https://git.sr.ht/~cadence/cloudtube/tree/main/item/pug/video.pug#L62)

there are also other software you can change YouTube rewrite URL (to another YouTube front-end or invidious-redirect) and it you can set invidious.redirect.io in config file (default, server) or preferences (user, stored with cookies)
- [nitter](https://github.com/zedeus/nitter)

some websites use invidious-redirect (redirect.invidious.io) as YouTube URL rewrite:
- [nitter.pussthecat.org](https://github.com/PussTheCat-org/PussTheCat.org-Configs/blob/master/Services/Nitter/nitter.conf#L26)

if you want **actual content**:
- https://nitter.pussthecat.org/YouTube
- https://nitter.pussthecat.org/YouTube/status/1360363141947944964#m
- https://tube.cadence.moe/watch?v=BaW_jenozKc (look for the invidious link)
- …

**Pull request merged**